Annotated __new__, binary_symbols, and as_set methods with type hints.

MayankBalyan opened the pull request:
#### Description of changes

Added type annotations to `sympy/sets/contains.py` as part of the ongoing typing effort.
- Annotated `__new__`, `binary_symbols`, and `as_set`.
- Added `from __future__ import annotations`.
- Handled circular imports for `Set` using `TYPE_CHECKING` and local imports inside `__new__`.

#### Issue number

Issue: https://github.com/sympy/sympy/issues/28806

#### References to other Issues or PRs

#### Brief description of what is fixed or changed

#### Other comments

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* sets
  * Added type hints to `contains.py`
<!-- END RELEASE NOTES -->

MayankBalyan committed:
Annotated __new__, binary_symbols, and as_set methods with type hints.

#### Description of changes

Added type annotations to `sympy/sets/contains.py` as part of the ongoing typing effort.
- Annotated `__new__`, `binary_symbols`, and `as_set`.
- Added `from __future__ import annotations`.
- Handled circular imports for `Set` using `TYPE_CHECKING` and local imports inside `__new__`.

#### Issue number

Issue: https://github.com/sympy/sympy/issues/28806

MayankBalyan committed:
Add type ignores for mypy compatibility in Contains class

oscarbenjamin commented:
Rather than using `type: ignore` we need to tell the type checker the type of the args. I think the proper fix should be like:
```diff
diff --git a/sympy/sets/contains.py b/sympy/sets/contains.py
index 403d487527..4834c188df 100644
--- a/sympy/sets/contains.py
+++ b/sympy/sets/contains.py
@@ -1,6 +1,7 @@
+from typing import TYPE_CHECKING
 from sympy.core import S
+from sympy.core.basic import Basic
 from sympy.core.sympify import sympify
-from sympy.core.relational import Eq, Ne
 from sympy.core.parameters import global_parameters
 from sympy.logic.boolalg import Boolean
 from sympy.utilities.misc import func_name
@@ -28,7 +29,13 @@ class Contains(Boolean):
 
     .. [1] https://en.wikipedia.org/wiki/Element_%28mathematics%29
     """
-    def __new__(cls, x, s, evaluate=None):
+
+    if TYPE_CHECKING:
+        @property
+        def args(self) -> tuple[Basic, Set]:
+            ...
+
+    def __new__(cls, x: Basic, s: Set, evaluate: bool | None = None) -> Boolean:  # type: ignore[misc]
         x = sympify(x)
         s = sympify(s)
 
@@ -53,11 +60,9 @@ def __new__(cls, x, s, evaluate=None):
         return super().__new__(cls, x, s)
 
     @property
-    def binary_symbols(self):
-        return set().union(*[i.binary_symbols
-            for i in self.args[1].args
-            if i.is_Boolean or i.is_Symbol or
-            isinstance(i, (Eq, Ne))])
+    def binary_symbols(self) -> set[Basic]:
+        bool_args = [a for a in self.args[1].args if isinstance(a, Boolean)]
+        return set().union(*[i.binary_symbols for i in bool_args])
 
-    def as_set(self):
+    def as_set(self) -> Set:
         return self.args[1]
```
The `type: ignore` on `__new__` is needed because of a mypy bug:
https://github.com/python/mypy/issues/15182

MayankBalyan committed:
Refactor args typing and binary_symbols logic.

Applied @oscarbenjamin suggestion to define args type in TYPE_CHECKING block and filter Booleans in binary_symbols, removing the need for type ignores.

MayankBalyan committed:
Refactor import of Set in contains.py

Removed unnecessary import of Set inside the Contains class and updated the import statement to be at the top of the file.

MayankBalyan commented:
Done! I have moved the Set import to the top level and removed the local import inside __new__ as requested.

MayankBalyan committed:
Remove unnecessary blank line in contains.py

MayankBalyan commented on sympy/sets/contains.py:
```diff
@@ -1,10 +1,14 @@
+
 
 
 class Contains(Boolean):
```
My apologies, I missed that spacing rule. I've corrected it to ensure there are exactly two blank lines.
